engine/table/src/main/java/io/deephaven/engine/table/impl/asofjoin/TypedAsOfJoinFactory.java

```diff
@@ -0,0 +1,409 @@
+import io.deephaven.chunk.LongChunk;
+import io.deephaven.engine.rowset.RowSet;
+import io.deephaven.engine.rowset.RowSetFactory;
+import io.deephaven.engine.rowset.WritableRowSet;
+import io.deephaven.engine.rowset.chunkattributes.OrderedRowKeys;
+import io.deephaven.engine.table.impl.NaturalJoinModifiedSlotTracker;
+import io.deephaven.engine.table.impl.by.alternatingcolumnsource.AlternatingColumnSource;
+import io.deephaven.engine.table.impl.by.typed.HasherConfig;
+import io.deephaven.engine.table.impl.naturaljoin.DuplicateRightRowDecorationException;
+import io.deephaven.util.QueryConstants;
+import org.jetbrains.annotations.NotNull;
+
+public class TypedAsOfJoinFactory {
+    public static void staticBuildLeftFound(HasherConfig<?> hasherConfig, boolean alternate,
+            CodeBlock.Builder builder) {
+        builder.addStatement("addLeftIndex(tableLocation, rowKeyChunk.get(chunkPosition))");
+    }
+
+    public static void staticBuildLeftInsert(HasherConfig<?> hasherConfig, CodeBlock.Builder builder) {
+        builder.addStatement("addLeftIndex(tableLocation, rowKeyChunk.get(chunkPosition))");
+        builder.addStatement("rightRowSetSource.set(tableLocation, $T.builderSequential())", RowSetFactory.class);
+        builder.addStatement("hashSlots.set(hashSlotOffset.getAndIncrement(), (long)tableLocation)");
+    }
+
+    public static void staticBuildRightFound(HasherConfig<?> hasherConfig, boolean alternate,
+            CodeBlock.Builder builder) {
+        builder.addStatement("addRightIndex(tableLocation, rowKeyChunk.get(chunkPosition))");
+    }
+
+    public static void staticBuildRightInsert(HasherConfig<?> hasherConfig, CodeBlock.Builder builder) {
+        builder.addStatement("addRightIndex(tableLocation, rowKeyChunk.get(chunkPosition))");
+        builder.addStatement("hashSlots.set(hashSlotOffset.getAndIncrement(), (long)tableLocation)");
+    }
+
+    public static void staticProbeDecorateLeftFound(HasherConfig<?> hasherConfig, boolean alternate,
+            CodeBlock.Builder builder) {
+        builder.addStatement("final long indexKey = rowKeyChunk.get(chunkPosition)");
+        builder.beginControlFlow("if (addLeftIndex(tableLocation, indexKey) && hashSlots != null)");
+        builder.addStatement("hashSlots.set(hashSlotOffset.getAndIncrement(), tableLocation)");
+        builder.addStatement("foundBuilder.addKey(indexKey)");
+        builder.endControlFlow();
+    }
+
+    public static void staticProbeDecorateLeftMissing(CodeBlock.Builder builder) {
+        // builder.addStatement("leftRedirections.set(hashSlotOffset++, $T.NULL_ROW_KEY)", RowSet.class);
+    }
+
+    public static void staticProbeDecorateRightFound(HasherConfig<?> hasherConfig, boolean alternate,
+            CodeBlock.Builder builder) {
+        builder.addStatement("addRightIndex(tableLocation, rowKeyChunk.get(chunkPosition))");
+    }
+
+    public static void rightIncrementalRehashSetup(CodeBlock.Builder builder) {
```

Comment:
The rest of this is cruft; but since you are going to do incremental maybe you wanted to leave it?